ydb/public/lib/ydb_cli/import/import.cpp

```diff
@@ -245,42 +266,70 @@ TStatus TImportFileClient::Import(const TVector<TString>& filePaths, const TStri
         case EOutputFormat::JsonBase64:
         case EOutputFormat::Parquet:
             break;
-        default: 
+        default:
             return MakeStatus(EStatus::BAD_REQUEST,
                         TStringBuilder() << "Unsupported format #" << (int) settings.Format_);
     }
 
+    bool isStdoutInteractive = IsStdoutInteractive();
+    size_t filePathsSize = filePaths.size();
+    std::mutex progressWriteLock;
+    std::atomic<size_t> globalProgress{0};
+
+    auto writeProgress = [&]() {
+        auto globalProgressValue = globalProgress.load();
+        std::lock_guard<std::mutex> lock(progressWriteLock);
+        Cout << "Progress " << (globalProgressValue / filePathsSize) << '%' << "\r";
+        Cout.Flush();
+    };
+
+    auto start = TInstant::Now();
+
+    auto pool = CreateThreadPool(filePathsSize);
+    TVector<NThreading::TFuture<TStatus>> asyncResults;
+
+    // If the single empty filename passed, read from stdin, else from the file
+
     for (const auto& filePath : filePaths) {
-        auto func = [this, &filePath, &dbPath, &settings] {
+        auto func = [&, this] {
+            std::unique_ptr<TFileInput> fileInput;
+            std::optional<i64> fileSizeHint;
+
             if (!filePath.empty()) {
                 const TFsPath dataFile(filePath);
+
                 if (!dataFile.Exists()) {
                     return MakeStatus(EStatus::BAD_REQUEST,
                         TStringBuilder() << "File does not exist: " << filePath);
                 }
+
                 if (!dataFile.IsFile()) {
                     return MakeStatus(EStatus::BAD_REQUEST,
                         TStringBuilder() << "Not a file: " << filePath);
                 }
-            }
 
-            if (settings.Format_ == EOutputFormat::Tsv) {
-                if (settings.Delimiter_ != "\t") {
-                    return MakeStatus(EStatus::BAD_REQUEST,
-                        TStringBuilder() << "Illegal delimiter for TSV format, only tab is allowed");
+                TFile file(filePath, OpenExisting | RdOnly | Seq);
+                i64 fileLength = file.GetLength();
+                if (fileLength) {
+                    fileSizeHint = fileLength;
                 }
-            }
 
-            auto result = NDump::DescribePath(*SchemeClient, dbPath);
-            auto resultStatus = result.GetStatus();
+                fileInput = std::make_unique<TFileInput>(file, settings.FileBufferSize_);
+            }
 
-            if (resultStatus != EStatus::SUCCESS) {
-                return MakeStatus(EStatus::SCHEME_ERROR,
-                    TStringBuilder() <<  result.GetIssues().ToString() << dbPath);
+            ProgressCallbackFunc progressCallback;
+
+            if (isStdoutInteractive)
+            {
+                size_t oldProgress = 0;
```

Comment:
Do we have guaranty that oldProgress allocated on stack would not be reused out of if () scope?

Reply:
We catch `oldProgress` in lambda by value, and then pass this lambda by value into upsert function.